### PR TITLE
Add spark3-hudi image

### PR DESCRIPTION
**Pull request by codope**
Add spark3-hudi image for product tests

- Installs Spark 3.2.1
- Installs hudi-spark3-bundle
- Add Hive user

### CODE REVIEW

**Comment by ebyhr** [comment]
@codope Is this PR ready for review?

**Comment by codope** [comment]
Not yet. I'll update it by tomorrow.

**Comment by codope** [comment]
@ebyhr @findinpath I have updated the PR. When I try to build locally using `make testing/spark3-hudi`, I get this error:
```
Dockerfile:13
--------------------
  11 |     # limitations under the License.
  12 |
  13 | >>> FROM testing/centos7-oj11:unlabelled
  14 |
  15 |     ARG SPARK_VERSION=3.2.1
--------------------
error: failed to solve: testing/centos7-oj11:unlabelled: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```
I also tried to first build the dependent centos first using `make testing/centos7-oj11` but still getting the same error.

File: etc/compose/spark3-hudi/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+version: '2.0'
+services:
+  spark:
+    image: testing/spark3-hudi:latest
```
Comment (nineinchnick) [inline etc/compose/spark3-hudi/docker-compose.yml:4]: Add a EOL at EOF

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -20,6 +20,7 @@ jobs:
             test: spark3-iceberg
           - image: spark3-delta
             test: spark3-delta
+          - image: spark3-hudi
```
Comment (nineinchnick) [inline .github/workflows/ci.yml:23]: Can you enable tests for this image? We just recently added them to almost all images. See `bin/test.sh`. The test is just a simple smoke test to see if a container using this image will start.

File: testing/spark3-hudi/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+EXPOSE 10213
+
+HEALTHCHECK --interval=10s --timeout=5s --start-period=10s \
+    CMD curl -f http://localhost:10213/
+CMD spark-submit \
+   --master "local[*]" \
+   --class org.apache.spark.sql.hive.thriftserver.HiveThriftServer2 \
+   --name "Thrift JDBC/ODBC Server" \
+   --conf spark.hive.server2.thrift.port=10213 \
+   spark-internal
```
Comment (nineinchnick) [inline testing/spark3-hudi/Dockerfile:53]: Add EOL at EOF

**Comment by findinpath** [comment]
@codope I just tried to build the image locally and the building process went just fine.
Make sure to rebase your changes on top of `master` before doing again `make testing/spark3-hudi`

File: testing/spark3-hudi/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM testing/centos7-oj11:unlabelled
+
+ARG SPARK_VERSION=3.2.1
```
Comment (findinpath) [inline testing/spark3-hudi/Dockerfile:15]: From

```


Hudi | Supported Spark 3 version
-- | --
0.12.x | 3.3.x (default build), 3.2.x, 3.1.x
0.11.x | 3.2.x (default build, Spark bundle only), 3.1.x
```

Given that hudi is a fresh connector in the Trino ecosystem, let's make use of the latest hudi `0.12.x` which has support for the latest Spark version `3.3.0` which can also execute on `Java 8/11/17` versions.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (codope) [reply .github/workflows/ci.yml:23]: done

**Comment by codope** [comment]
@nineinchnick @findinpath I still get the error after rebase. Here's the gist of full output of make command: https://gist.github.com/codope/921469ef910e3a76314d251b35060a6e

**Comment by codope** [comment]
> Can you also include the output of `docker images | grep testing`? It looks like the `docker tag` command runs fine, but subsequent `docker buildx` commands can't see the new image built locally. Can you also check if you're not running out of disk space?

@nineinchnick Thanks for the pointers. Looks like i'm running out of space. I have just 0.5gb left. I'll clean up and retry in some time. How much space is required typically?
Output of `docker info` in: https://gist.github.com/codope/921469ef910e3a76314d251b35060a6e
Output of `docker images | grep testing` below
```
docker images | grep testing
ghcr.io/trinodb/testing/hdp3.1-hive                         65                             90bbaa186970   2 months ago    4.5GB
ghcr.io/trinodb/testing/hdp2.6-hive                         65                             953bc3fc663d   2 months ago    3.23GB
testing/spark3-hudi                                         latest                         76ec389112ca   2 months ago    1.37GB
testing/centos7-oj11                                        latest                         e84dea2f6000   2 months ago    966MB
testing/centos7-oj11                                        unlabelled                     e84dea2f6000   2 months ago    966MB
ghcr.io/trinodb/testing/hdp3.1-hive                         64                             4b99d1995fbd   3 months ago    4.5GB
ghcr.io/trinodb/testing/spark3.0-iceberg                    64                             cd45c86b997d   3 months ago    1.33GB
ghcr.io/trinodb/testing/centos7-oj11                        64                             e822b8bf2137   3 months ago    966MB
```

**Comment by codope** [comment]
I pruned all older images, removed the build cache, and reclaimed enough space. This time it's a different error:
`Error response from daemon: No such image: testing/centos7-oj11:latest`. Full log below:

<details>
<summary><b>make testing/spark3-hudi</b></summary>

```
bin/depend.sh -p unlabelled testing/spark3-hudi/Dockerfile testing/kerberos:unlabelled testing/cdh5.12-hive-kerberized:unlabelled testing/cdh5.15-hive-kerberized:unlabelled testing/accumulo:unlabelled testing/centos7-oj11:unlabelled testing/centos7-oj17-openldap-referrals:unlabelled testing/centos7-oj17:unlabelled testing/hdp2.6-hive-kerberized-2:unlabelled testing/gpdb-6:unlabelled testing/spark3-iceberg:unlabelled testing/hdp2.6-hive:unlabelled testing/cdh5.15-hive-kerberized-kms:unlabelled testing/hive3.1-hive:unlabelled testing/spark3-hudi:unlabelled testing/phoenix5:unlabelled testing/centos7-oj17-openldap:unlabelled testing/hdp3.1-hive:unlabelled testing/cdh5.12-hive:unlabelled testing/cdh5.15-hive:unlabelled testing/dns:unlabelled testing/hdp3.1-hive-kerberized:unlabelled testing/hdp2.6-hive-kerberized:unlabelled testing/spark3-delta:unlabelled
testing/centos7-oj11:unlabelled
bin/depend.sh -p unlabelled testing/centos7-oj11/Dockerfile testing/kerberos:unlabelled testing/cdh5.12-hive-kerberized:unlabelled testing/cdh5.15-hive-kerberized:unlabelled testing/accumulo:unlabelled testing/centos7-oj11:unlabelled testing/centos7-oj17-openldap-referrals:unlabelled testing/centos7-oj17:unlabelled testing/hdp2.6-hive-kerberized-2:unlabelled testing/gpdb-6:unlabelled testing/spark3-iceberg:unlabelled testing/hdp2.6-hive:unlabelled testing/cdh5.15-hive-kerberized-kms:unlabelled testing/hive3.1-hive:unlabelled testing/spark3-hudi:unlabelled testing/phoenix5:unlabelled testing/centos7-oj17-openldap:unlabelled testing/hdp3.1-hive:unlabelled testing/cdh5.12-hive:unlabelled testing/cdh5.15-hive:unlabelled testing/dns:unlabelled testing/hdp3.1-hive-kerberized:unlabelled testing/hdp2.6-hive-kerberized:unlabelled testing/spark3-delta:unlabelled
docker pull library/centos:7
7: Pulling from library/centos
Digest: sha256:c73f515d06b0fa07bb18d8202035e739a494ce760aa73129f60f4bf2bd22b407
Status: Image is up to date for centos:7
docker.io/library/centos:7

Building [testing/centos7-oj11@latest] image using buildkit

cd testing/centos7-oj11 && time /bin/sh -c "( docker buildx build --compress --progress=plain --add-host hadoop-master:127.0.0.2   -t testing/centos7-oj11:latest --label io.trino.git.hash=f231f96dc51f183d7b87a6ce7b7da27f1efa5e02 . )"
WARNING: No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 1.81kB done
#1 DONE 0.2s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.2s

#3 [internal] load metadata for docker.io/library/centos:7
#3 DONE 8.6s

#4 [internal] load build context
#4 DONE 0.0s

#5 [1/3] FROM docker.io/library/centos:7@sha256:c73f515d06b0fa07bb18d8202035e739a494ce760aa73129f60f4bf2bd22b407
#5 resolve docker.io/library/centos:7@sha256:c73f515d06b0fa07bb18d8202035e739a494ce760aa73129f60f4bf2bd22b407
#5 resolve docker.io/library/centos:7@sha256:c73f515d06b0fa07bb18d8202035e739a494ce760aa73129f60f4bf2bd22b407 5.2s done
#5 DONE 5.7s

#4 [internal] load build context
#4 transferring context: 385B done
#4 DONE 0.5s

#6 [2/3] COPY ./files /
#6 CACHED

#7 [3/3] RUN     set -xeu &&     yum install -y         nc         wget         &&         rpm -i https://cdn.azul.com/zulu$([ "$(arch)" != "aarch64" ] || echo "-embedded")/bin/zulu11.58.23-ca-jdk11.0.16.1-linux."$(arch)".rpm &&     alternatives --set java /usr/lib/jvm/zulu-11/bin/java &&     alternatives --set javac /usr/lib/jvm/zulu-11/bin/javac &&         yum --enablerepo=extras install -y setuptools epel-release &&     yum install -y python-pip &&     pip install supervisor &&         yum install -y         less `# helpful when troubleshooting product tests`         net-tools `# netstat is required by run_on_docker.sh`         sudo         telnet `# helpful when troubleshooting product tests`         vim `# helpful when troubleshooting product tests`         jq `# helpful json processing tool`         &&     yum -y clean all && rm -rf /tmp/* /var/tmp/*
#7 CACHED

real	0m15.684s
user	0m0.209s
sys	0m0.186s
docker history testing/centos7-oj11:latest
Error response from daemon: No such image: testing/centos7-oj11:latest
make: *** [testing/centos7-oj11@latest] Error 1
```

</details>

**Comment by nineinchnick** [comment]
The warning there looks related. Your docker server version looks like the latest one, but your buildx plugin is not. Can you check if you could update it?

**Comment by codope** [comment]
I am still facing the same issue after upgrading buildx. Let me check with others. Maybe it has got something to do with my local environment. Just to ensure I'm doing it right, there is no step other than running `make testing/spark3-hudi` for this PR right?

**Comment by codope** [comment]
Also, I see that the spark3-hudi image was built and smoke test ran successfully: https://github.com/trinodb/docker-images/actions/runs/3235524333/jobs/5300075791
Can we merge this PR? @nineinchnick

**Comment by nineinchnick** [comment]
Yes, but we have to wait for a maintainer to do so

**Comment by nineinchnick** [comment]
> Just to ensure I'm doing it right, there is no step other than running `make testing/spark3-hudi` for this PR right?

That's right, there are no other steps required. I saw some issues reported in Docker that buildx does not load the image into the local repo after building it, but it is supposed to be fixed. If you have the latest veesion, as us, you should not be affected. 

Building the image locally is only required if you want to test it to run product tests in Trino, which i assume you do want, since you're working on adding new tests.

You could try building the images without using make, it prints all the docker commands it executes.

File: testing/spark3-hudi/Dockerfile (hunk 4 above)
Comment (ebyhr) [reply testing/spark3-hudi/Dockerfile:15]: @codope Hudi doesn't support different version between server and client?

**Comment by codope** [comment]
Thanks for the help @nineinchnick .
Actually, my colleague can build it with the make command. So, it's my local environment that is causing some issues. I am going to uninstall docker and start from scratch. 
For product tests, once the spark3-hudi image is released, I can pull and use it (even if I can't build it locally).
@ebyhr @findinpath This PR is ready to land now. Would be great if you can help with releasing the image.

**Comment by findinpath** [comment]
@codope can you please add a basic product test creating a table via spark hudi and reading it via trino in https://github.com/trinodb/trino/issues/14365 with the newly created image so that we can actually verify that the image indeed basically works? Otherwise, we may end up that something is missing and we need to do yet another release.

**Comment by findinpath** [comment]
@codope  https://github.com/trinodb/trino/pull/14669 draft PR  is a working draft which you can use as a sample template for building the tests required for https://github.com/trinodb/trino/issues/14365

I based the tests (same as for Delta Lake OSS) on a MinIO data lake because I think (@codope  correct me if I'm wrong) this kind of setup is close to the integration with AWS S3 which we want to test with Trino for Hudi.

@ebyhr please do test the draft PR on your machine as well and if you're ok with the results, I think we can release a new version of the docker images so that @codope  can continue the work on the product tests.

**Comment by findinpath** [comment]
@codope pls squash the commits and add a relevant commit message.

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
I'm testing the draft PR based on this. Let me merge once it's complete.

**Comment by findinpath** [comment]
@ebyhr it was a copy-paste issue in `SuiteHudi` - corrected in the meantime. `EnvSinglenodeDeltaLakeOss` -> `EnvSinglenodeHudi`

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Merged, thanks!

**Pull request merged**